glommio/src/sync/semaphore.rs

```diff
@@ -7,49 +7,115 @@ use std::cell::RefCell;
 use std::future::Future;
 use std::io::{Error, ErrorKind, Result};
 use std::pin::Pin;
-use std::rc::Rc;
 use std::task::{Context, Poll, Waker};
 
-use intrusive_collections::intrusive_adapter;
+use intrusive_collections::linked_list::LinkOps;
+use intrusive_collections::{container_of, offset_of, Adapter, PointerOps};
 use intrusive_collections::{LinkedList, LinkedListLink};
+use std::marker::PhantomPinned;
+use std::ptr::NonNull;
 
 #[derive(Debug)]
-struct Waiter {
-    node: Rc<WaiterNode>,
-    sem_state: Rc<RefCell<SemaphoreState>>,
+struct Waiter<'a> {
+    node: WaiterNode,
+    semaphore: &'a Semaphore,
 }
 
 #[derive(Debug)]
 struct WaiterNode {
     link: LinkedListLink,
-    state: RefCell<WaiterState>,
+    units: u64,
+    waker: RefCell<Option<Waker>>,
+
+    //waiter node can not be Unpin so its pointer could be used inside of intrusive
+    //collection, it also can not outlive the container which is guaranteed by the
+    //Waiter lifetime bound to the Semaphore which is container of all Waiters.
+    _p: PhantomPinned,
 }
 
-#[derive(Debug)]
-struct WaiterState {
-    units: u64,
-    waker: Option<Waker>,
+struct WaiterPointerOps;
+
+unsafe impl PointerOps for WaiterPointerOps {
+    type Value = WaiterNode;
+    type Pointer = NonNull<WaiterNode>;
+
+    unsafe fn from_raw(&self, value: *const Self::Value) -> Self::Pointer {
+        NonNull::new_unchecked(value as *mut Self::Value)
+    }
+
+    fn into_raw(&self, ptr: Self::Pointer) -> *const Self::Value {
+        ptr.as_ptr() as *const Self::Value
```

Comment:
why isn't a non-null check needed here?

It's true that this looks just like a cast to me, but at the same time it has a self parameter, which indicates that the user of this data structure may expect a transformation to happen. If a transformation could happen, could it be that a null pointer is ever passed?

Keep in mind that I am just being too paranoid with this but I want to understand the unsafe usage better

Reply:
Is it because we define `NonNull` as our pointer type?

Reply:
Hi @glommer , in their terminology, Pointer is what you pass into the intrusive collection. In our case that is NonNull, so we do not pass a raw pointer there but valid object.

Reply:
For example https://github.com/Amanieu/intrusive-rs/blob/master/src/linked_list.rs#L1041 push method accepts pointer. 
Also in PoiterOps we declared Pointer https://github.com/DataDog/glommio/pull/228/files/46aa06bd3f1f9cf978d2fdf94645591c168ae0ff#diff-9063a13a75caca6f9b4f70da0df9795d578d990677ae6c5ec0f916a43181d6c6R40 as NonNull defining all subsequent operations of LinkedList as such which will accept NonNull.